oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/ClientAuthorization.java

```diff
@@ -0,0 +1,203 @@
+                                                          clientCredentials.getValue());
+        } else {
+            // we should not get here
+            throw new NullPointerException("authorizationSupplier && credentialsSupplier");
+        }
+        return clientAuthorization;
+    }
+
+    /**
+     * Fetches client authorization token or client credentials from the supplier and composes client
+     * {@code Authorization} header value,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>:
+     * <pre>{@code
+     * Authorization: Basic czZCaGRSa3F0Mzo3RmpmcDBaQnIxS3REUmJuZlZkbUl3
+     * }</pre>.
+     *
+     * @return encoded client {@code Authorization} header value.
+     */
+    public String authorizationHeaderValue() {
+        return CaseUtil.firstUpperAllLowerCase(authorizationType) +
+               AUTHORIZATION_SEPARATOR + composeAuthorizationString();
+    }
+
+    /**
+     * Fetches client credentials from the supplier and composes required body parameters,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>:
+     * <pre>{@code
+     * client_id=s6BhdRkqt3&client_secret=7Fjfp0ZBr1KtDRbnfVdmIw
+     * }</pre>.
+     * The client MAY omit the {@code client_secret} parameter if the client secret is an empty string.
+     */
+    public void setCredentialsAsBodyParameters(QueryParamsBuilder formBuilder) {
+        requireNonNull(credentialsSupplier, "credentialsSupplier");
+        final Map.Entry<String, String> clientCredentials = credentialsSupplier.get();
+        formBuilder.add(CLIENT_ID, requireNonNull(clientCredentials.getKey(), CLIENT_ID));
+        final String clientSecret = clientCredentials.getValue();
+        if (clientSecret != null && !clientSecret.isEmpty()) {
+            formBuilder.add(CLIENT_SECRET, clientSecret);
+        }
+    }
+
+    /**
+     * Fetches client credentials from the supplier and composes required body parameters,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>:
+     * <pre>{@code
+     * client_id=s6BhdRkqt3&client_secret=7Fjfp0ZBr1KtDRbnfVdmIw
+     * }</pre>.
+     * The client MAY omit the {@code client_secret} parameter if the client secret is an empty string.
+     *
+     * @return encoded client credentials request body parameters as a {@link String}.
+     */
+    public String credentialsBodyParameters() {
+        final QueryParamsBuilder formBuilder = QueryParams.builder();
+        setCredentialsAsBodyParameters(formBuilder);
+        return formBuilder.toQueryString();
+    }
+
+    private static String encodeClientCredentials(String clientId, String clientSecret) {
+        return Base64.getEncoder()
+                     .encodeToString(
+                             (clientId + CREDENTIALS_SEPARATOR + clientSecret)
+                                     .getBytes(StandardCharsets.UTF_8));
+    }
+}
```

Comment:
As you can see, this class `ClientAuthorization` does not possess any actual credentials. It only holds a reference to the the entire authorization string `Supplier` or to credentials `Supplier`. This is done intentionally, so this class does not cache and maintain the lifecycle of the credentials/authorization. Credentials may change over time, credentials (or authorization string) should not be exposed. Therefore, there is nothing to print in the 'toString()' method besides the authorization type (`Basic`, `Bearer`, etc). Do you think it still makes sense to print just the type?
We should not invoke `toString()` on either of the suppliers, as it's opaque to our framework and we don't know whether it will expose the credentials or not.

Reply:
printing only the type of the authorization and the type of the source, like:
{"type": "Basic", "source": "credentials"}

Reply:
Sounds good to me. Thanks for elaboration, @max904-github